src/components/EditableField/EditableFieldComposite.tsx

```diff
@@ -0,0 +1,301 @@
+          if (name.includes(m.name)) {
+            return { ...m, text: value[0].value }
+          }
+          return m
+        }),
+      })
+    }
+  }
+
+  handleEnter = passedFn => {
+    return () => {
+      passedFn && passedFn()
+
+      this.setState({ inputState: null, hasActiveFields: false }, () => {
+        this.maskRef.setAttribute('tabindex', '0')
+        this.maskRef.focus()
+      })
+    }
+  }
+
+  handleEscape = passedFn => {
+    return ({ value, name }) => {
+      const { maskItems } = this.state
+      passedFn && passedFn()
+
+      this.setState(
+        {
+          inputState: null,
+          hasActiveFields: false,
+          maskItems: maskItems.map(m => {
+            if (name.includes(m.name)) {
+              return { ...m, text: value[0].value }
+            }
+            return m
+          }),
+        },
+        () => {
+          this.maskRef.setAttribute('tabindex', '0')
+          this.maskRef.focus()
+        }
+      )
+    }
+  }
+
+  handleMaskClick = name => {
+    if (name === 'placeholder') {
+      const input = this.groupRef.querySelector('input')
+
+      input && input.focus()
+      return
+    }
+
+    const inputs = this.groupRef.querySelectorAll('input')
+
+    // It is tested
+    /* istanbul ignore next */
+    if (inputs) {
+      for (let index = 0; index < inputs.length; index++) {
+        const element = inputs[index]
+
+        if (element.id.includes(name)) {
+          element.setSelectionRange &&
+            element.setSelectionRange(
+              element.value.length,
+              element.value.length
+            )
+          element.focus()
+          return
+        }
+      }
+    }
+  }
+
+  handleMaskKeyDown = event => {
+    const isEnter = event.key === key.ENTER
+    const isEscape = event.key === key.ESCAPE
+
+    if (isEnter) {
+      const input = this.groupRef.querySelector('input')
+
+      input && input.focus()
+      this.maskRef.removeAttribute('tabindex')
+    }
+    // It is tested
+    /* istanbul ignore next */
+    else if (isEscape) {
+      this.maskRef.removeAttribute('tabindex')
+    }
+  }
+
+  renderMaskContent = () => {
+    const { placeholder, separator } = this.props
+    const { maskItems } = this.state
+
+    const maskItemsWithValue = maskItems.filter(m => Boolean(m.text))
+
+    if (maskItemsWithValue.length > 0) {
+      return maskItemsWithValue.map((m, index, self) => (
+        <span
+          className={COMPOSITE_CLASSNAMES.maskItem}
+          key={m.name}
+          onClick={() => {
+            this.handleMaskClick(m.name)
+          }}
+        >
+          {m.text}
+          {index !== self.length - 1 ? `${separator}\u00a0` : ''}
+        </span>
+      ))
+    }
+
+    return (
+      <span
+        className={classNames(
+          COMPOSITE_CLASSNAMES.maskItem,
+          STATES_CLASSNAMES.isPlaceholder
+        )}
+        onClick={() => {
+          this.handleMaskClick('placeholder')
```

Comment:
having the onClick created like that, would that not declare a lot of unnecessary functions in memory each time masks are re-rendered? (same with line 243).

I get that we need to pass custom arguments based on which mask was clicked. Not sure I have a solution for that too...maybe caching each callback in an object could help 🤷‍♂

Reply:
Hi PL, welcome back!

This subject of inline functions in React is a bit touchy, there is a bit of a dogma that you should just NEVER do it. But most of the time this is said without any evidence. The reason people give is that react has to create a new function every time `render` runs and thus it is BAD for performance.

I believe in staying away from premature micro optimizations. Creating and destroying functions in JavaScript is _fast_, arrow functions are native and/or transpiled with babel to normal functions. Before, the pattern was to use `bind`, and that _was_ slow (although I think JS engines have make it less so).

In any case, if we find a problem and after measuring and finding this to be the culprit, we address it at that point. Right now, it will require some kind of convoluted code just to avoid it, and don't think is worth investing the time.

I like this article by Ryan Florence: https://cdb.reacttraining.com/react-inline-functions-and-performance-bdff784f5578, which puts more eloquently what my thoughts are.

In searching for the article, is amazing the number of articles that start with "ANTIPATTERN", "NEVER DO THIS", hahahaha  ¯\_(ツ)_/¯

It's definitely an interesting subject and we can talk more about it!

Reply:
thank you for your input, I will read the article for sure. I agree that micro optimizing that kind of stuff is trivial and that we should do it only if it's required